Release/v.1.0.0

FiniDG opened the pull request:
Release first production version (V1.0.0)

Unverified commits:
- ready for server deployment + readme update v.1.0.0
- added except for fullgnomen upload error
- small adjustments before release, mainly column ordering]

Verified commits:
Login functionality #6 + #8 
- Made the "data pages" only accessible to logged users and made Admin pages
- added Logging

Cleanup layout #9 
- implemented the PEP8 rules
- added and deleted some comments for clarity
- removed the unused function get_all_fields()
- reformatted HTML
- removed unused navbar functions (with defaults that were inactivated)

Alissa connection instead of using importing Excels from "O-schijf" #10 
- adds a LastUpdatedOn in MongoDB with a default (for first time use script).
- looks into Alissa based on LastUpdatedOn, to minimize the number of analyses to go trough and only update if necessary
- checks if analysis is already in MongoDB and updates it if necessary
- removed the old O-schijf (Excel) upload method
- changed the logging to "INFO", because otherwise cluttered with Alissa HTTP attempts (and busy responses)
- added config (empty in GitHub) for Alissa connection keys and login info

Layout changes to accommodate for the new Alissa method (replacing the Excel upload method) #13 
- changed layout of the variant, gene, all and patient pages.
- changed the layout and information found on the "VUS" page.
- added GnomAD link functionality. Now only available for SNP. Other SNVs will be added in the future after some Alissa updates.
- removed "false" links that did not lead to a specific page or anything related to what was clicked on. At first everything was clickable, now only the functional.
- formatting of code + removal of old commented code

Added restrictions for specific roles and updated account page #15 
- new roles ROLE_CLINIC and ROLE_DIAGGEN added
- make restrictions for ROLE_CLINIC that they can never access the "Genes", "Variants" and "List All" pages
- make restrictions for ROLE_CLINIC that they cannot link via another page to "Genes", "Variants" and "List All" pages
- updated the "Account" page to let the user know about these new restrictions (you are always redirected to this page if access is denied)
- restrictions also apply for users that are not ROLE_DIAGGEN or ROLE_ADMIN

ellendejong commented on extra/import_data.py:
```diff
@@ -0,0 +1,174 @@
+    patient["platform_dataset"] = analyis_sources["platformDataSet"]["info"]
+
+    # changes the format of the annotation sources from Alissa, to fit in MongoDB and resemble the old O-schijf format
+    externalSources_dict = {}
```
Compared to #10 , i miss line `variants = vus_export`
Should this be added here or later on line 124 maybe?

FiniDG replied:
I saw during my last work that this line was completely useless. I did not use it in that code block (I think it was a remnant of earlier trial and errors). 
later there used to be a code line `for variant in variants:`, and that is now changed to `for variant in vus_export:`, so nothing has changed content-wise

ellendejong commented on extra/import_data.py:
```diff
@@ -0,0 +1,174 @@
+    for source in analyis_sources['annotationSources']:
+        source_name = ""
+        source_value = ""
+        for key, value in source.items():
+            if key == 'name':
+                source_name = value
+            elif key == 'description':
+                source_value = value
```
What happens when there are multiple annotationSources?
Seems only the last annotationSource is set to `source_name` and `source_value`, while others will be ignored. Is that correct?

FiniDG replied:
there are multiple annotation sources `for source in analyis_sources['annotationSources']:`, but all of them need to be rearranged , which is what the other loop does. so for each annotation source, the values are reset by using         `source_name = ""` and `source_value = ""`.

ellendejong commented on extra/import_data.py:
```diff
@@ -0,0 +1,174 @@
+    for variant in vus_export:
+        # format for gnomad Links. 4 availble in Alissa (snp, insertion, deletion and substitution)
+        try:
+            fullgnomen = variant['platformDatasets']['HGVS genomic-level nomenclature (fullGNomen)']
+            variant['fullgnomen'] = fullgnomen  # NC_000001.10:g.12345678T>A
+            if fullgnomen:  # NC_000001.10:g.12345678T>A
+                gnomad_data = re.split(':[a-z].', fullgnomen)[1]  # 123456789T>A
+                if variant["type"] == "snp":
+                    gnomad_data = [c for c in re.split(r'([-+]?\d*\.\d+|\d+)', gnomad_data) if c]  # 123456789T>A
+                    gnomad_data = variant["chromosome"] + "-" + re.sub('[<>]+', '-', ("-".join(gnomad_data)))  # 1-12345678-T-A
+                    variant['GnomadVariant'] = {'Single nucleotide variant': gnomad_data}
+                elif variant["type"] in ["insertion", "deletion", "substitution"]:
+                    gnomad_data = variant["chromosome"] + "-" + gnomad_data
+                    variant['GnomadVariant'] = {variant["type"].capitalize(): gnomad_data}
+                    # TODO: make link format correctly for "insertion", "deletion", "substitution" genomic variations
+            else:  # on rare occasions, fullGNomen is empty
+                variant['GnomadVariant'] = {variant["type"]: ''}
```
In my opinion, this try block is too big and not specific enough. Is the keyError and TypeError likely to happen in each line?

FiniDG replied:
yes and both are due to incomplete entries in Alissa. So there is "technically" a VUS, but it is a very old entry that does not actually contain any data hence the error message.

ellendejong replied:
Is it possible to check whether data is complete before extracting and checking all other conditions? In that way the try becomes more specific and as a benefit also smaller.

FiniDG replied:
there are several checks in place to check for complete data. most of them (should be all) can be checked if retrieved from Alissa. However, this error only occurred ~5 times on the whole Alissa database. The person in charge already made a ticket about these weird entries in Alissa.

So technically, yes I could check beforehand, but then I should check every key-value in Alissa (which I think is not feasable)

FiniDG replied:
We talked about this, and one of the issues is also a too long try-except.
would this be a solution?
```
    for variant in vus_export:
        # format for gnomad Links. 4 availble in Alissa (snp, insertion, deletion and substitution)
        try:
            fullgnomen = variant['platformDatasets']['HGVS genomic-level nomenclature (fullGNomen)']
            variant['fullgnomen'] = fullgnomen  # NC_000001.10:g.12345678T>A
        except (KeyError, TypeError):
            logger.error('Variant in patient %s has no platformDatasets and fullGNomen, variant not uploaded' % patient_dn_no)
            continue
        if fullgnomen:  # NC_000001.10:g.12345678T>A
            gnomad_data = re.split(':[a-z].', fullgnomen)[1]  # 123456789T>A
            if variant["type"] == "snp":
            ......
```

FiniDG replied:
I have tested above solution without "typeError" to see what exactly causes this error (which line mainly).

```
2022-06-03 20:35:08,769 [INFO] __main__ - start Alissa retrieval of: DN201803750 with analysisID: 33569
2022-06-03 20:35:19,333 [INFO] __main__ - Alissa retrieval completed of: DN201803750
2022-06-03 20:35:19,342 [INFO] __main__ - start uploading to mongodb: DN201803750
Traceback (most recent call last):
  File "import_data.py", line 169, in <module>
    import_from_alissa(alissa_client, start_time, logger)
  File "import_data.py", line 69, in import_from_alissa
    upload_to_mongodb(inheritance_analysis, accession_number, analyis_sources, patient_dn_no, vus_export, logger)
  File "import_data.py", line 128, in upload_to_mongodb
    fullgnomen = variant['platformDatasets']['HGVS genomic-level nomenclature (fullGNomen)']
TypeError: string indices must be integers
```

So I think that if the key is wrong/non-existant you get the "keyError", for the "typeError" it happens at the same line, when the index is non-existant [stackoverflowexample](https://stackoverflow.com/questions/6077675/why-am-i-seeing-typeerror-string-indices-must-be-integers)

The variable item is a string. An index looks like this:
```
mystring = 'helloworld'
print mystring[0]
'h'
```
The above example uses the 0 index of the string to refer to the first character.
Strings can't have string indices (like dictionaries can). So this won't work:
```
mystring = 'helloworld'
print mystring['stringindex']
TypeError: string indices must be integers
```

So in this case, I think my code example will work for both the typeError and the keyError

ellendejong replied:
Right. You could split this into two try-except, one based on `variant['platformDatasets']` and another on the second key. However, might be my personal preference.

ellendejong commented on extra/import_data.py:
```diff
@@ -0,0 +1,174 @@
+                elif variant["type"] in ["insertion", "deletion", "substitution"]:
+                    gnomad_data = variant["chromosome"] + "-" + gnomad_data
+                    variant['GnomadVariant'] = {variant["type"].capitalize(): gnomad_data}
+                    # TODO: make link format correctly for "insertion", "deletion", "substitution" genomic variations
```
In my opinion, those kind of TODO comments should not be part of a v1.0.0 release.

FiniDG replied:
see earlier comments

ellendejong commented on src/datatable.py:
```diff
@@ -0,0 +1,192 @@
+        },
+        :return: filter dict
+        """
+        # TODO : search per individual column? not needed?
```
In my opinion, those kind of TODO comments should not be part of a v1.0.0 release.

FiniDG replied:
👍 (not mine, but I agree)

ellendejong commented on extra/logging_config.yml:
```diff
@@ -0,0 +1,22 @@
+    filename: /data/vusualizer/VUSualizer/extra/output.log
+root:
+  level: DEBUG
+  handlers: [console, file]
```
Based on #8, i expected `email` here as well.

FiniDG replied:
I did remove "email" in the handlers, because I did not use it. I just finished some small tests by adding the handler to act on "logging.error", but I cannot get it to work properly right now.
I already made an git issue #11 , based on another pull request. I think we use a different way to send emails within our team, so I rather want to have this implemented in a future release.

ellendejong commented on run.py:
```diff
@@ -0,0 +1,26 @@
+import sys
+
+# /data/vusualizer/VUSualizer, use location of VUSualizer on the serverside
+sys.path.insert(0, '/data/vusualizer/VUSualizer')
```
Could this be configurable? Either via a config file or parameter parsing?
Seems to be the only difference between `run.py` and `run_local.py`. Please avoid code duplications.

ellendejong commented on src/config.py:
```diff
@@ -0,0 +1,23 @@
+'''
+Used by __init__.py to set the MongoDB configuration for the whole app
+The app can then be references/imported by other script using the __init__.py
```
typo: references vs referenced

FiniDG replied:
👍

ellendejong commented on src/config.py:
```diff
@@ -0,0 +1,23 @@
+TESTING = True
+
+# mongo db
+SECRET_KEY = "vusualyzerrrrrrrr"
```
Should a secret key be part of git?

rernst replied:
See: https://github.com/UMCUGenetics/ExonCov/blob/9fab496a41c6e28ff22e5a193ea32735d39373bf/config.py#L16 

I randomly generate one for each server we checkout ExonCov on.

rernst replied:
So indeed your production key should not be on github.

FiniDG replied:
We talked about this in teams. I searched (partly) wrong. It is a key from flask [link](https://flask.palletsprojects.com/en/2.1.x/config/#SECRET_KEY).
Should not be on Github. I will change this to `SECRET_KEY = ""` and give an explanation in the Readme to change this. if not filled in by someone, it will generate an error.

If someone changes the key it is not a big deal. cookie sessions will reset and people need to login again.
The secret key can be used to generate fake cookies, to gain access. therefore I will remove it from Github and make sure that the key is always something random.

ellendejong reviewed (changes requested):
Review part 1.

Since I don't want to lose my pending comments, I decided to submit multiple reviews.

ellendejong commented on src/datatable.py:
```diff
@@ -0,0 +1,192 @@
+        pipeline = pipeline + [{"$sort": sorting}, {"$skip": pages.start}]
+        if pages.length >= 0:
+            pipeline.append({"$limit": pages.length})
+        self.result_data = list(self.dbh[self.collection].aggregate(pipeline))
```
With my (very limited) knowledge regarding classes and OOP,  I believe it is common and good practice to use getters and setters. I am a bit worried about changing class specific variables in different functions. What are your thoughts about this?

FiniDG replied:
see earlier comment:
I have made another issue #19, so that the next version needs to have a look.

FiniDG replied:
I think al the "getting" and "setting" is done using the datatables plugin (javascripts). [link](https://datatables.net/)
I think that the datainput is regulated by the datatables.py by sending queries to the mongoDB database. this is then returned to the datatables plugin to use as variables ect. This way it is always dynamic and every datatable can be made on the fly, and not all information is stored in a python dictionary.

ellendejong commented on src/datatable.py:
```diff
@@ -0,0 +1,192 @@
+                    val = "-"
+
+                if isinstance(val, dict):
+                    print("YAS")
+                    val = json.dumps(val)
+                    print(val)
```
Are the print on line 48 and 50 being used? Or were those part of the development/debugging phase?

FiniDG replied:
removed them. I have checked, and it will print those statements to the commandline. I guess it was needed to check when that function would be used when clicking though the VUSualizer. So I indeed believe it was once part of an debugging phase.

ellendejong commented on src/datatable.py:
```diff
@@ -0,0 +1,192 @@
+        output['data'] = data_rows
+        return output
+
+    def run_queries(self):
```
When is this function called / executed? For any type of query or are there specific queries or a generic deliverable?
Based on the code, I get the impression it is more of a filter and grouping function than performing a query. 
I wonder if the function title is specific enough and matches with its functionality.

FiniDG replied:
I think this code runs queries to the mongodb database, to retrieve information that will show in the datatable, and then filtered/sorted based on the settings provided.
So for example if I want to use the search bar it needs to change the totals.

The filtering is done with another function called `filtering()`. which is called as one of the first lines in `run_queries()`.

It is a generic deliverable, as any datatable in the VUSualizer is made based on the information that needs to be shown at that time. So although the function name is a bit vague, I also think that this is due to the broadness of this whole file (datatable.py). It takes information (for example dictionaries), but you do not know beforehand what it is, because it is a very generic script that is used multiple times for all datatables.

I left the name as is, but if you have a good suggestion I can change it.

ellendejong commented on src/datatable.py:
```diff
@@ -0,0 +1,192 @@
+
+    def filtering(self):
+        """
+        build your filter spec
+        "search": {
+            "regex": "false",
+            "value": ""
+        },
```
Could you explain these lines? I don't understand these comments. I cannot find any variable with search, neither as input to the function or output.

FiniDG replied:
see earlier comment:
I have made another issue #19, so that the next version needs to have a look.

FiniDG replied:
I think this was left by the developer to help understand the difference between python and mongodb. But because this can be looked up in the mongodb documentation I removed them.

ellendejong commented on src/datatable.py:
```diff
@@ -0,0 +1,192 @@
+            # need to match for every field
+            # the term put into search is logically concatenated with 'or' between all columns
+            or_filter_on_all_columns = []
+            for k, v in self.request["columns"].items():
```
Are the variables `k` and `v`  short-lived variables or loop-index variables? Or, could the variable have a more meaningful name?

FiniDG replied:
see earlier comment:
I have made another issue #19, so that the next version needs to have a look.

FiniDG replied:
I also tried to explain this a bit more in an earlier comment.
`datatables.py` is a very generic script. Any datatable in the VUSualizer is made based on the information that needs to be shown at that time. It takes information (for example dictionaries), but you do not know beforehand what it is. are these genes? variants? patients? or something else you search for in the search bar?
I think that is why the developer choose to use variable names like "key" and "value", because it is a dictionary with unknown content that is being called.
key and value, or k and v, are accepted names for these dictionary variables within the python community.

ellendejong commented on src/datatable.py:
```diff
@@ -0,0 +1,192 @@
+
+    def sorting(self):
+        """
+        mongo translation for sorting order
+        "order": {
+            "0": {
+                "column": "0",
+                "dir": "asc"
+            }
+        },
+        :return: order dict { col name : direction }
```
Could you explain these lines?

ellendejong commented on src/forms.py:
```diff
@@ -1,37 +1,42 @@
         InputRequired(),
         Email(message='Invalid email'),
-        Length(max=30, message='Your email is too long')
+        Length(max=60)
```
Message not needed anymore?

FiniDG replied:
not deleted by me (previous programmer), good that you spotted it

ellendejong commented on src/forms.py:
```diff
@@ -1,37 +1,42 @@
     password = PasswordField('password', [
         InputRequired(),
-        Length(min=8, max=20, message='Your password is either too short or too long.')
+        Length(min=8, max=30, message='Your password is either too short or too long.')
```
Should this match with password requirements of the registration? Line nr 15.

FiniDG replied:
👍

ellendejong commented on src/static/css/sb-admin-2.css:
```diff
@@ -0,0 +1,10998 @@
+.was-validated .form-control:valid, .form-control.is-valid {
+  border-color: #1cc88a;
+  padding-right: calc(1.5em + 0.75rem);
+  background-image: url("data:image/svg+xml,%3csvg xmlns='http://www.w3.org/2000/svg' viewBox='0 0 8 8'%3e%3cpath fill='%231cc88a' d='M2.3 6.73L.6 4.53c-.4-1.04.46-1.4 1.1-.8l1.1 1.4 3.4-3.8c.6-.63 1.6-.27 1.2.7l-4 4.6c-.43.5-.8.4-1.1.1z'/%3e%3c/svg%3e");
```
Based on [Google CSS style code,](https://google.github.io/styleguide/htmlcssguide.html) i think it should be `https` instead of `http`.

FiniDG replied:
I had a closer look at the three CSS files. These are standardized files, see for example this link for this specific file

[link](https://github.com/StartBootstrap/startbootstrap-sb-admin-2/blob/master/css/sb-admin-2.css)

ellendejong replied:
In that case: I wonder if this file is custom developed by 'us', or whether css should move to `statis/vendor/`

rernst replied:
So, 'vendor' is typically used to show that 'static' files are from external origin. CSS or JS (or Fonts) does not really matter. For example this is completely fine: https://github.com/UMCUGenetics/VUSualizer/tree/dev/src/static/vendor/bootstrap -> contains JS and CSS.

I do think you have jquery twice in the static folder -> https://github.com/UMCUGenetics/VUSualizer/tree/dev/src/static/js and https://github.com/UMCUGenetics/VUSualizer/tree/dev/src/static/vendor/jquery

ellendejong reviewed (changes requested):
part 2

ellendejong commented on src/templates/index.html:
```diff
@@ -0,0 +1,136 @@
+<!-- End of Footer -->
+
+
+{% endblock %}
```
I see a warning: "No newline at end of file".

FiniDG replied:
not needed for HTML

ellendejong commented on src/templates/genes.html:
```diff
@@ -0,0 +1,43 @@
+</div>
+<!-- /.container-fluid -->
+
+{% endblock %}
```
I see a warning: "No newline at end of file".

FiniDG replied:
not needed for HTML

ellendejong commented on src/views/main.py:
```diff
@@ -0,0 +1,188 @@
+import json
+
+
+# make connection with MongoDB data
+variant_col = mongo.db.variant
+user_col = mongo.db.user
+
+columns = ['#', '_id', 'total']
+default_fields = ["Details", "dn_no", "gene", "Position", "inheritanceMode", "cdna", "protein", "effect", "ref", 
+                  "genotype Patient", "genotype Mother", "genotype Father", "inheritedFrom", "GnomAD", "fullgnomen",
+                  "transcript"]
+default_order = {"dn_no": 1, "gene": 1}
+variants = []
```
Are these variables required to be global variables?
And/or should they move to a `if __name__ == "__main__"` block

FiniDG replied:
it is used by multiple html files, basically all  the main pages (genes, patients, variants), so I think it is needed like this

ellendejong commented on src/models.py:
```diff
@@ -0,0 +1,43 @@
+import datetime
+import uuid
+import json
```
./src/models.py:1:1: F401 'datetime' imported but unused
./src/models.py:2:1: F401 'uuid' imported but unused
./src/models.py:3:1: F401 'json' imported but unused

FiniDG replied:
👍

ellendejong commented on src/models.py:
```diff
@@ -0,0 +1,43 @@
+import datetime
+import uuid
+import json
+from flask_login import UserMixin
```
./src/models.py:5:1: F811 redefinition of unused 'UserMixin' from line 4

ellendejong commented on src/__init__.py:
```diff
@@ -0,0 +1,25 @@
+admin = Admin(app, name='VUSualizer admin', template_mode='bootstrap3')
+
+
+from . import *
+from .views import main, auth, admin_view
```
flake8 detects a few issues regarding the import statement.

./src/__init__.py:24:1: E402 module level import not at top of file
./src/__init__.py:24:1: F403 'from . import *' used; unable to detect undefined names
./src/__init__.py:24:1: F401 'from . import *' imported but unused
./src/__init__.py:25:1: E402 module level import not at top of file
./src/__init__.py:25:1: F401 '.views.main' imported but unused
./src/__init__.py:25:1: F401 '.views.auth' imported but unused
./src/__init__.py:25:1: F401 '.views.admin_view' imported but unused

FiniDG replied:
removed the import * 👍 
however kept de .views at the end of the file instead of beginning. Need to be loaded after creating the flask app. these are also not "unused" but refer to the files in de "views" folder (main.py, auth.py and admin_view.py).

I added a comment to prevent this from happening again according to the flake8 documentation
[flake documentation](https://flake8.pycqa.org/en/stable/user/violations.html#in-line-ignoring-errors)

ellendejong commented on src/views/auth.py:
```diff
@@ -0,0 +1,57 @@
+from flask import render_template, request, redirect, url_for, flash
```
./src/views/auth.py:1:1: F401 'flask.request' imported but unused

FiniDG replied:
👍

ellendejong commented on src/views/admin_view.py:
```diff
@@ -0,0 +1,38 @@
+    password = fields.StringField('Password')
+    active = fields.BooleanField('Active')
+    role = fields.StringField('Role')
+
+# add UserView, where admin can change properties of users
+class UserView(ModelView):
+
```
./src/views/admin_view.py:18:1: E302 expected 2 blank lines, found 1

ellendejong commented on src/views/main.py:
```diff
@@ -0,0 +1,188 @@
+user_col = mongo.db.user
+
+columns = ['#', '_id', 'total']
+default_fields = ["Details", "dn_no", "gene", "Position", "inheritanceMode", "cdna", "protein", "effect", "ref", 
+                  "genotype Patient", "genotype Mother", "genotype Father", "inheritedFrom", "GnomAD", "fullgnomen",
+                  "transcript"]
```
./src/views/main.py:15:113: W291 trailing whitespace

FiniDG replied:
👍

ellendejong commented on src/views/main.py:
```diff
@@ -0,0 +1,188 @@
+variants = []
+
+
+def login_required(f):
+    '''Wrapper around the login_required() function of Flask, to also check the active status of the user'''
+    @wraps(f)
+    def decorated_function(*args, **kwargs):
+        if current_user.is_authenticated is False or check_if_user_active(current_user):
+            return redirect(url_for('account'))
+        return f(*args, **kwargs)
+    return decorated_function
```
This function name is the same as one of your imports from flask_login. Either rename the function or use import `as` statement

FiniDG replied:
it is meant to be the same name with the @wraps. you change the imported function

ellendejong reviewed (changes requested):
Happy to see this pull-request towards the first release of the VUSualizer, great!
I have a few general remarks.

### external plugins to vendor
Robert explained to me that the vendor folder is often used for external software / plugins and should be ignored when reviewing the code.
He pointed out that `src/static/js/jquery-3.4.1.min.js` is also part of `src/static/vendor/jquery/jquery-3.4.1.min.js`. 
It would help if all code developed by you/us/our team is part of static/js or static/css and all external code should go to static/vendor. 

- [ ] Could you have a look into the src/static/js and src/static/css folders and move files when it concerns external software?

### import_data.py
Based on previous pull-requests, I get the idea that this script has grown over time quite a bit.

- [ ] Personally, I would prefer to see more bite sized functions, and using `return` more often in import_data.py

For example:
- retrieve_info_analysis_alissa
- check_valid_analysis 
- retrieve_vus_gus_analysis
- upload_to_mongodb


### naming conventions
Some variables use a very generic and meaningless naming convention. 

- [ ] Maybe you could go over the code changes and consider if/where variable names can be improved?

### CSS files
Some of the .css files are huge and I am worried about sustainability as well as readibility.
I would suggest to have a look into CSS coding style guides, for example https://google.github.io/styleguide/htmlcssguide.html
There are a few CSS lint extensions available. Please use these as well. 
I know we discussed CSS previously and decided it was out of scope and to postpone it to next release.
I think it would improve a lot if you could add comments of the different sections/group of code in each .css file, for example src/static/css/sb-admin-2.css could have more comments. 

- [ ] Add comments to custom .css files describing code blocks.

### Remaining notes
- [x] For each html in src/templates, I see a warning message: "No newline at end of file". Could you perform some HTML lint checking and see whether newline should be there?
- [ ] I added a few comments related to flake8 output. Might be good to rerun flake8 just before merge to master branch.
- [ ]

FiniDG commented on .gitignore:
```diff
@@ -2,3 +2,9 @@ __pycache__
 .idea
 *.xml
 venv/
+extra/input/*
+extra/verwerkt/*
+extra/output.log
```
It is where the upload scripts to the mongoDB are. used to be for the excel to MongoDB upload, but now it uses Alissa

FiniDG commented on .gitignore:
```diff
@@ -2,3 +2,9 @@ __pycache__
+extra/input/*
+extra/verwerkt/*
+extra/output.log
+.DS_Store
+extra/.DS_Store
```
👍

FiniDG commented on .gitignore:
```diff
@@ -2,3 +2,9 @@ __pycache__
+extra/output.log
+.DS_Store
+extra/.DS_Store
+src/__init__.pyc
```
👍

FiniDG commented on README.md:
```diff
@@ -0,0 +1,100 @@
+.
+|──────run.py - used to run the flask app, also doubles as .wsgi file
+|──────run_local.py - used to run the flask app on your local machine
+|──────src/
```
I have explained the "extra" folder more, but the others are defaults used for webpage building and flask.
https://python-adv-web-apps.readthedocs.io/en/latest/flask3.html

FiniDG commented on README.md:
````diff
@@ -0,0 +1,100 @@
+$ source ./venv/bin/activate
+$ pip install -r requirements.txt
+```
+2. Install MongoDB
````
👍 , although I am not sure if this one is also used on the UMCU servers, but it was used on my local computer

FiniDG commented on README.md:
```diff
@@ -0,0 +1,100 @@
+- /_get_gene_data
+- /_get_patient_data
+
+These urls are handled in main.py. The program then uses the DataTablesServer class from datatable.py to query the database and return the data back to datatables.js to create a table that can be clicked through. 
```
this is an explanation on how the previous programmer set up the flask module. I just left it as is

FiniDG commented on README.md:
````diff
@@ -0,0 +1,100 @@
+```
+
+## Run for development
+Activate venv
+```
+$ python run.py
+```
````
made it more clear. it were 2 seperate steps

FiniDG commented on extra/import_data.py:
```diff
@@ -0,0 +1,174 @@
+import pymongo
+import logging
+import json
+import logging.config
```
was needed seperately. see lines 162 and 163 where they are used both

FiniDG commented on extra/import_data.py:
```diff
@@ -0,0 +1,174 @@
+        analysis_report = alissa_client.get_analysis_report(analysis['id'])
+        if analysis['classificationTreeName'] and analysis_report:
+            # retrieve basic info from Alissa about the analysis
+            # patient_dn_no = alissa_client.get_analysis_report(analysis['id'])[0]['reportName'].split("_")[0]
```
👍

FiniDG commented on extra/import_data.py:
```diff
@@ -0,0 +1,174 @@
+                except json.decoder.JSONDecodeError:
+                    break
+            logger.info('Alissa retrieval completed of: %s' % patient_dn_no)
+            # sometimes the there are no VUS marked or found within an analysis, then no info needs to be uploaded
```
👍

FiniDG commented on extra/import_data.py:
```diff
@@ -0,0 +1,174 @@
+                logger.info('Patient %s, has no VUS marked/found. Not uploaded to MongoDB' % patient_dn_no)
+            elif vus_export is None:
+                logger.info('Patient %s not uploaded, Alissa database temporarily not available' % patient_dn_no)
+                # TODO send (email) notification of this error, see issue #11 GitHub
```
removed the issue number, however I would like to keep the TODO

FiniDG commented on extra/import_data.py:
```diff
@@ -0,0 +1,174 @@
+            logger.info('removed: %s from database, start replacing with newer version' % patient_dn_no)
+        elif last_updated_on_Alissa < last_updated_on_mongoDB:
+            logger.info('lastUpdatedOn older than within the database for %s, this should not happen' % patient_dn_no)
+            # TODO send (email) notification of this error, see issue #11 GitHub
```
see previous comment

FiniDG commented on extra/import_data.py:
```diff
@@ -0,0 +1,174 @@
+        elif last_updated_on_Alissa > last_updated_on_mongoDB:
+            db.delete_many({"dn_no": patient_dn_no})
+            logger.info('removed: %s from database, start replacing with newer version' % patient_dn_no)
+        elif last_updated_on_Alissa < last_updated_on_mongoDB:
+            logger.info('lastUpdatedOn older than within the database for %s, this should not happen' % patient_dn_no)
```
This should not happen. When this triggers the data in the VUSualizer is newer then in Alissa. I wanted to add this line, otherwise it will not report anything. The VUSualizer is a new app, so when this issue occurs, it is good to find out what exactly triggered it and we need to discuss with research/lab on how to deal with it or prevent this. However, since I do not know it this is theoretically possible now, I at least want to log it somewhere.
I hope this answers the question?

ellendejong replied:
Sounds like a serious issue if the elif would be triggered. Is it possible to change logger.info to a more serious level, such as error?

FiniDG replied:
yes! It used to trigger the wrong handler, but now that is removed, so it should work. Will change it 👍

FiniDG commented on extra/import_data.py:
```diff
@@ -0,0 +1,174 @@
+    # get all relevant info from one patient into one dictionary.
+    patient = {}
+    patient["dn_no"] = patient_dn_no
+    logger.info('start uploading to mongodb: %s' % patient_dn_no)
```
this code clock is about uploading the data to mongodb (not the specific line). line 148 is when it is finished

ellendejong replied:
This conversation is resolved, because of code preference.

FiniDG commented on extra/import_data.py:
```diff
@@ -0,0 +1,174 @@
+            # TODO send (email) notification of this error, see issue #11 GitHub
+
+    # get all relevant info from one patient into one dictionary.
+    patient = {}
+    patient["dn_no"] = patient_dn_no
+    logger.info('start uploading to mongodb: %s' % patient_dn_no)
+    patient["patient_accession_no"] = accession_number
+    patient.update(inheritance_analysis)
+    patient["analysis"] = patient.pop("reference")
+    patient["platform_dataset"] = analyis_sources["platformDataSet"]["info"]
```
This is simply code preference. I like this step by step and also easy to add new steps. this is also due to multiple sources of the data (some info is only found in the inheritance data, others in the VUS export data ect.

ellendejong replied:
This conversation is resolved, because of code preference.

FiniDG commented on README.md:
````diff
@@ -0,0 +1,100 @@
+```
+2. Install MongoDB
+3. Load data into MongoDB with script in the "extra" folder.
+4. (Optional) Configurate MONGO_URI in config.py if you're using different user details and/or database name
````
👍

FiniDG commented on README.md:
````diff
@@ -0,0 +1,100 @@
+$ pip install -r requirements.txt
+```
+2. Install MongoDB
+3. Load data into MongoDB with script in the "extra" folder.
````
👍

FiniDG commented on src/datatable.py:
```diff
@@ -0,0 +1,192 @@
+            i += 1
+            data_row.append("<a href='/vus/{0}'>{1}</a>".format(row['_id'], i))
+            for col in self.columns:
+                if col in row:
+                    val = row[col]
+                else:
+                    val = "-"
+
+                if isinstance(val, dict):
+                    print("YAS")
+                    val = json.dumps(val)
+                    print(val)
+                elif isinstance(val, list):
+                    val = ", ".join(val)
+
+                if isinstance(val, str):
+                    val = (val[:max_string_length] + '...') if len(val) > max_string_length else val
```
I think it is a valid question, however one I cannot answer at this moment.
I do agree that this comment you make is within the scope of this pull request, however I haven't touched or worked this file when developing the new features mentioned in the text I wrote for this pull request.

reasoning:
I haven't anticipated for this to happen. To explain a little bit, the previous programmer (who also left this project completely) did almost everything in the development branch, because there was not a first production version yet. Only some test versions. I had not noticed this on time, so when I opened this release branch to merge everything into the main branch, these code changes were presented to you as a reviewer.

I do want to leave these comments for now, because:
1. This code was validated earlier (see all commits in this pull request by previous programmer)
2. The first production version is now well on its way
3. The current code works correctly
4. I expect a lot of changes based on the 1st production version in scope and functionalities. To have a in-depth look at it now seems a bit overkill.

I have made another issue in github #19, so that the next version needs to have a look.

ellendejong replied:
Yes, you are right.  The `if isinstrance(val, str)` is still needed in case `row[col]` is a string.
The `if isinstrance(val, str)` will not be used on `val = "-"`. Since the only thing that happens inside the if statement is stripping the value if it is too long. "-" will never be too long, so the check and stripping is not required.

FiniDG commented on src/datatable.py:
```diff
@@ -0,0 +1,192 @@
+        self.run_queries()
+
+    def output_result_on_given_fields(self):
+        output = {}
+        output['draw'] = str(int(self.request['draw']))
+        output['recordsTotal'] = str(self.records_total)
+        output['recordsFiltered'] = str(self.records_filtered)
```
see earlier comment:
I have made another issue #19, so that the next version needs to have a look.

ellendejong replied:
This conversation is resolved, because of code preference.

FiniDG commented on src/datatable.py:
```diff
@@ -0,0 +1,192 @@
+        output['data'] = data_rows
+        return output
+
+    def output_result_on_queried_fields(self):
```
see earlier comment:
I have made another issue #19, so that the next version needs to have a look.

ellendejong replied:
Second question regarding return value of the function is personal preference.

The first question whether `output_result_on_given_fields` and `output_result_on_queried_fields` are code duplications is still valid. Let me know if you agree that it is a code duplication.

FiniDG replied:
I was busy with finding out what the difference was. The functions seem to have a lot in common, but they were also very different (if you compared them next to each other).
I tried to find out what each function does exactly. the most used function is "given_fields" and the other does not seem to be used at all. I think I know why.

Before, all text was a link in the VUSualizer. I changed this in my own feature-branches to only gene, variant and patient. The reasoning was that all other links "did not work" properly. They referenced back to the "list_all" page, but did not seem to do anything there. I suspect that the intention was:
1. I click on "chromosome 5" *you can think of any text here
2. go to the "list all" page and show every variant that is also on chromosome 5
Also seems to explain the filtering going on and the name of the function (queried fields). I removed the entire code here and also the code in main.py. The functionality is not used and did not function 100% yey, so if this needs to be implemented in the future it might be better to start from scratch.

FiniDG commented on src/datatable.py:
```diff
@@ -0,0 +1,192 @@
+            data_row.append(i)
+            for col, val in row.items():
+                uwu = ""
+                if col == "_id":
+                    if self.group_by == "fullgnomen":
+                        uwu = '<a href="/variant/{0}">{0}</a>'.format(val)
+                    elif self.group_by == "dn_no":
+                        uwu = '<a href="/patient/{0}">{0}</a>'.format(val)
+                    elif self.group_by == "gene":
+                        uwu = '<a href="/gene/{0}">{0}</a>'.format(val)
```
see earlier comment:
I have made another issue #19, so that the next version needs to have a look.

FiniDG replied:
this is solved in an earlier comment, this function is now removed

FiniDG commented on src/datatable.py:
```diff
@@ -0,0 +1,192 @@
+            group["$group"]["protein"] = {"$addToSet": "$pnomen"}
+        pipeline = []
+
+        if group:
+            pipeline.append(group)
+        if filter:
+            pipeline.append(match)
+        pipeline = pipeline + [{"$sort": sorting}, {"$skip": pages.start}]
+        if pages.length >= 0:
+            pipeline.append({"$limit": pages.length})
+        self.result_data = list(self.dbh[self.collection].aggregate(pipeline))
+
+        if group:
+            # total amount unfiltered & total amount filtered (search bar)
+            self.records_total = len(list(self.dbh[self.collection].aggregate([group])))
+            self.records_filtered = len(list(self.dbh[self.collection].aggregate([group, match])))
+        else:
+            self.records_total = len(list(self.dbh[self.collection].find()))
+            self.records_filtered = len(list(self.dbh[self.collection].find(filter)))
```
see earlier comment:
I have made another issue #19, so that the next version needs to have a look.

ellendejong replied:
This conversation is resolved, because of code preference.

FiniDG commented:
I think I addressed most of your points. I await your next review and possible comments on my reply's
Things I did not address:
- overhaul of functions in import_data.py
- some comments were in files that I are outside the scope of this pull request. see #19 and [link to reply](https://github.com/UMCUGenetics/VUSualizer/pull/17#discussion_r885813699)

ellendejong commented on extra/import_data.py:
```diff
@@ -0,0 +1,173 @@
+        if source_name and source_value:
+            externalSources_dict[source_name] = source_value
+        else:
+            logger.info("error, externalsources has changed format in Alissa")
```
Is it possible to change the logger level to error?

FiniDG replied:
yes I already did this, based on your earlier comment about this 👍

FiniDG commented on README.md:
````diff
@@ -59,14 +59,16 @@ $ python3 -m venv venv
 $ brew services start mongodb-community@4.2
 ```
-3. (Optional) Configurate MONGO_URI in config.py if you're using different user details and/or database name
+3. Configure the config.py (in src folder) and full in a random string for the SECRET_KEY. The app will not work withou a SECRET_KEY. A random string can be generated by using for example [String Generator](https://www.random.org/strings/) or any other string generator. Do not share this code to others, as this could potentially be used to generate fake cookies for logins. Changing it will result ...
````
👍 , I was not awake yet i guess

FiniDG commented:
We talked on teams about how to tackle these two functions in datatable.py. We agreed upon to make it one function.
After merging it into one function, we talked about how this could be further improved, because the initial code seems very similar.
After some research (added as comments to this function) the data imported in the function is quite different. The genes/variants/patients page mainly use the mongoDB collection index, while the other specifies specific columns and retrieve those entries from the mongoDB.

While trying to merge this (without having specific `if field is "given" or "queried"` loop), it became apparent that is would  make the function more complicated and less "followable/unstandable". At least that is my opinion. you then have to loop based on specific mongo collection columns or else given columns, which desired a different way of looping. 

Mongo entries require to query them using keys (and then appropriate extra quering if that key has multiple values in a dict or list)
Mongo collection entries are already queried.

Could you have another look at this proposed solution with extra comments and explanation? of course I would like to hear it if you see this differently.

hwmvandeutekom reviewed (approved):
Small changes, all approved
